File: flux-flix-service/src/main/java/com/example/fluxflixservice/FunctionalReactiveConfig.java
package com.example.fluxflixservice;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.MediaType;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;

import static org.springframework.web.reactive.function.server.RequestPredicates.*;
import static org.springframework.web.reactive.function.server.RouterFunctions.*;

/*
    Spring WebFlux enables functional definition of reactive routing
    function/routes
 */
//@Configuration
public class FunctionalReactiveConfig {
    @Bean
    RouterFunction<?> routerFunction(MovieService movieService) {
        return route(
                        GET("/movies"), request -> ServerResponse
                                .ok()
                                .body(movieService.getAllMovies(), Movie.class))
                .andRoute(
                        GET("/movies/{id}"), request -> ServerResponse
                                .ok()
                                .body(movieService.getMovieById(request.pathVariable("id")), Movie.class))
                .andRoute(
                        GET("/movies/{id}/events"), request -> ServerResponse
                                .ok()
                                .contentType(MediaType.TEXT_EVENT_STREAM)
                                .body(movieService.getEventsForMovie(request.pathVariable("id")), MovieEvent.class));
    }
}

File: flux-flix-service-kotlin/src/main/kotlin/com/example/fluxflixservicekotlin/FunctionalReactiveConfig.kt
package com.example.fluxflixservicekotlin

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.MediaType
import org.springframework.web.reactive.function.server.*

@Configuration
class FunctionalReactiveConfig(private val movieHandler: MovieHandler) {
    @Bean
    fun router() = router {
        "/movies".nest {
            accept(MediaType.APPLICATION_JSON).nest {
                GET("/", { request -> movieHandler.getMovies(request) })
            }
            accept(MediaType.APPLICATION_JSON).nest {
                GET("/{id}", { request -> movieHandler.getMovie(request) })
            }
            accept(MediaType.TEXT_EVENT_STREAM).nest {
                GET("/{id}/events", { request -> movieHandler.getEvents(request) })
            }
        }
    }


//    @Bean
//    fun routerFunction(movieHandler: MovieHandler): RouterFunction<ServerResponse> {
//        return router {
//            ("/movies").nest {
//                GET("") {
//                    serverRequest -> ServerResponse
//                        .ok()
//                        .body(movieHandler.getMovies(serverRequest))
//                }
//                GET("/{id}") {
//                    serverRequest ->  ServerResponse
//                        .ok()
//                        .body(movieHandler.getMovie(serverRequest))
//                }
//                GET("/{id}/events") {
//                    serverRequest -> ServerResponse
//                        .ok()
//                        .contentType(MediaType.TEXT_EVENT_STREAM)
//                        .body(movieHandler.getEvents(serverRequest))
//                }
//            }
//        }
//    }
}